### DISCUSSION TITLE
Type for ORM class?

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#8429] [Usage Questions] [answered]
Is there a base class or anything for ORM classes in sqlalchemy 2? If not, how would you type such a class for mypy?

The only thing I can find online, is this awkward function to check the type at runtime:
```
def is_sa_model(cls: Type[object]) -> bool:
    try:
        class_mapper(cls)
    except UnmappedClassError:
        return False
    return True
```

But, if I need to type the parameter of a function as accepting an ORM model, currently all I have is `Type[Any]`.

In other words, what I'd like to see is something more like this:
```
def do_something_with_model(model: sa.ORMType):
    assert isinstance(model, sa.ORMType)  # or issubclass() maybe
    ...
```

**Comment by zzzeek** [comment, marked as answer]
> Is there a base class or anything for ORM classes in sqlalchemy 2? If not, how would you type such a class for mypy?

im not sure this is what you're asking since it's on practically every page of the docs and you would have seen it, but SQLA 2 bases Declarative mappings on a class called [DeclarativeBase](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html?highlight=declarativebase#sqlalchemy.orm.DeclarativeBase).   This does not include classes that are "classically" mapped or classes that are mapped with decorators, but that's still a base class that can be passed around for typing even in those cases.

That said, if this is for your own application, it's even better to have your own Base that indicates classes that are "mapped", but maybe you're building some kind generic ORM tool, not sure.


> 
> The only thing I can find online, is this awkward function to check the type at runtime:
> 
> ```
> def is_sa_model(cls: Type[object]) -> bool:
>     try:
>         class_mapper(cls)
>     except UnmappedClassError:
>         return False
>     return True
> ```

that's a runtime check, how does that apply to mypy ?   are you trying to make a [TypeGuard](https://peps.python.org/pep-0647/)?  you'd need to return `TypeGuard[DeclarativeBase]` for mypy to recognize it at the typing level.

A good runtime check would be:

```py
def is_mapped(cls: Type[Any]) -> TypeGuard[DeclarativeBase]:
    insp = inspect(cls, raiseerr=False)
    return insp is not None and getattr(insp, 'is_mapper', False)
```

above, you can have mapped classes that aren't actually DeclarativeBase if you are using the decorator such as `@registry.mapped`, but the `DeclarativeBase` will still give it "ORM-like" behaviors.

**Reply by Dreamsorcerer** [reply]
Thanks, so `Type[DeclarativeBase]` should do the trick then. Yes, it's a generic tool (rebuilding aiohttp-admin), so I want to introspect arbitrary models from users.
